Importing the dependencies:

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
#loading the dataset into the pandas dataframe
credit_card_df = pd.read_csv('/content/drive/MyDrive/Datasets/creditcard.csv.zip', compression ='zip')

In [3]:
#checking the first five rows of the dataset

credit_card_df.head()

Time        V1        V2        V3  ...       V27       V28  Amount  Class
0   0.0 -1.359807 -0.072781  2.536347  ...  0.133558 -0.021053  149.62      0
1   0.0  1.191857  0.266151  0.166480  ... -0.008983  0.014724    2.69      0
2   1.0 -1.358354 -1.340163  1.773209  ... -0.055353 -0.059752  378.66      0
3   1.0 -0.966272 -0.185226  1.792993  ...  0.062723  0.061458  123.50      0
4   2.0 -1.158233  0.877737  1.548718  ...  0.219422  0.215153   69.99      0

[5 rows x 31 columns]

In [4]:
#checking the shape of the dataset
credit_card_df.shape

(284807, 31)

In [5]:
#last five rows can also be checked

credit_card_df.tail()

Time         V1         V2  ...       V28  Amount  Class
284802  172786.0 -11.881118  10.071785  ...  0.823731    0.77      0
284803  172787.0  -0.732789  -0.055080  ... -0.053527   24.79      0
284804  172788.0   1.919565  -0.301254  ... -0.026561   67.88      0
284805  172788.0  -0.240440   0.530483  ...  0.104533   10.00      0
284806  172792.0  -0.533413  -0.189733  ...  0.013649  217.00      0

[5 rows x 31 columns]

In [6]:
#dataset information

credit_card_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [7]:
#checking the number of missing values in each column:

credit_card_df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

As we can see there is no missing values at all in the dataset. So it is a plus point for us. Else we will have to handle the missing values by dropping them or by replacing them with some other values, this method is known as imputation. 

In [8]:
credit_card_df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

As we can see the number of values of two cases in our target variable is too much different which states that our dataset is too imbalanced. So, we will have to handle this by using the method such as undersampling where we will take some random sample from our data and will try to make it balanced. So that our model can be trained in a generalised way. 

Here 

0----> Normal Transaction

1----> Fraudulent Transaction



In [9]:
#seperating the data for the analysis:

legit = credit_card_df[credit_card_df.Class == 0]
fraud = credit_card_df[credit_card_df.Class == 1]

In [10]:
legit.shape

(284315, 31)

In [11]:
fraud.shape

(492, 31)

In [12]:
#statistical measure about the data of a particular column

legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [13]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [14]:
#comparing the values of both sort of transactions using groupby

credit_card_df.groupby('Class').mean()

Time        V1        V2  ...       V27       V28      Amount
Class                                    ...                                
0      94838.202258  0.008258 -0.006271  ... -0.000295 -0.000131   88.291022
1      80746.806911 -4.771948  3.623778  ...  0.170575  0.075667  122.211321

[2 rows x 30 columns]

Under-Sampling: As the dataset is highly imbalanced so this is a technique to overcome this. In this, we will build sample dataset containing similar distribution of normal and fraudulent transactions. 

As the number of fraudulent transactions is 492, so we will build a sample dataset of normal transactions too with 492 transactions only. 

In [15]:
legit_sample = legit.sample(n=492)

In [16]:
legit_sample.shape

(492, 31)

In [17]:
#concatenating the two datasets:

new_df = pd.concat([legit_sample, fraud], axis = 0)

In [18]:
new_df.shape

(984, 31)

In [19]:
new_df.head()

Time        V1        V2  ...       V28  Amount  Class
17341    28641.0 -0.519725  0.991902  ...  0.124305    3.58      0
67184    52404.0 -0.642732  1.697539  ...  0.192684    1.00      0
55811    47136.0  0.983936  0.232877  ...  0.027590   35.24      0
270803  164261.0 -1.748173  1.853680  ... -0.118231   17.97      0
54773    46609.0  0.105541 -0.963791  ...  0.088733  427.29      0

[5 rows x 31 columns]

In [20]:
new_df.tail()

Time        V1        V2  ...       V28  Amount  Class
279863  169142.0 -1.927883  1.125653  ...  0.147968  390.00      1
280143  169347.0  1.378559  1.289381  ...  0.186637    0.76      1
280149  169351.0 -0.676143  1.126366  ...  0.194361   77.89      1
281144  169966.0 -3.113832  0.585864  ... -0.253700  245.00      1
281674  170348.0  1.991976  0.158476  ... -0.015309   42.53      1

[5 rows x 31 columns]

All the fraudulent transactions are concatenated on the top of normal transactions, which can be seen using tail function.

In [21]:
new_df['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

Now we can see, our dataset is fully balanced!

In [22]:
new_df.groupby('Class').mean()

Time        V1        V2  ...       V27       V28      Amount
Class                                    ...                                
0      93096.772358 -0.187780  0.052282  ... -0.027919  0.000294   83.579837
1      80746.806911 -4.771948  3.623778  ...  0.170575  0.075667  122.211321

[2 rows x 30 columns]

Now we can see by grouping the transaction by mean that there is no huge difference in between the values of normal and fraudulent transactions. 

Splitting the data into features and targets:

In [23]:
X = new_df.drop(columns = 'Class', axis = 1)
Y = new_df['Class']

In [24]:
print(X)

            Time        V1        V2  ...       V27       V28  Amount
17341    28641.0 -0.519725  0.991902  ...  0.282588  0.124305    3.58
67184    52404.0 -0.642732  1.697539  ...  0.306229  0.192684    1.00
55811    47136.0  0.983936  0.232877  ...  0.034453  0.027590   35.24
270803  164261.0 -1.748173  1.853680  ... -1.467507 -0.118231   17.97
54773    46609.0  0.105541 -0.963791  ...  0.000891  0.088733  427.29
...          ...       ...       ...  ...       ...       ...     ...
279863  169142.0 -1.927883  1.125653  ...  0.292680  0.147968  390.00
280143  169347.0  1.378559  1.289381  ...  0.389152  0.186637    0.76
280149  169351.0 -0.676143  1.126366  ...  0.385107  0.194361   77.89
281144  169966.0 -3.113832  0.585864  ...  0.884876 -0.253700  245.00
281674  170348.0  1.991976  0.158476  ...  0.002988 -0.015309   42.53

[984 rows x 30 columns]


In [25]:
print(Y)

17341     0
67184     0
55811     0
270803    0
54773     0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


Splitting the data into training data and testing data for training the model.

In [26]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, stratify=Y, random_state = 2)

In [27]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


Model Training using Logistic Regression:

In [28]:
model = LogisticRegression()

In [29]:
#training the model with training data

model.fit(X_train, Y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


LogisticRegression()

##Model building on imbalanced data:
Metric selection for heavily imbalanced data
As we have seen that the data is heavily imbalanced, where only 0.17% transctions are fraudulent, we should not consider Accuracy as a good measure for evaluating the model. Because in the case of all the datapoints return a particular class(1/0) irrespective of any prediction, still the model will result more than 99% Accuracy.

Hence, we have to measure the ROC-AUC score for fair evaluation of the model. The ROC curve is used to understand the strength of the model by evaluating the performance of the model at all the classification thresholds. The default threshold of 0.5 is not always the ideal threshold to find the best classification label of the test point. Because the ROC curve is measured at all thresholds, the best threshold would be one at which the TPR is high and FPR is low, i.e., misclassifications are low. After determining the optimal threshold, we can calculate the F1 score of the classifier to measure the precision and recall at the selected threshold.

###Why SVM was not tried for model building and Random Forest was not tried for few cases?
In the dataset we have 284807 datapoints and in the case of Oversampling we would have even more number of datapoints. SVM is not very efficient with large number of datapoints beacuse it takes lot of computational power and resources to make the transformation. When we perform the cross validation with K-Fold for hyperparameter tuning, it takes lot of computational resources and it is very time consuming. Hence, because of the unavailablity of the required resources and time SVM was not tried.

**For the same reason Random forest was also not tried for model building in few of the hyperparameter tuning for oversampling technique.**

###Why KNN was not used for model building?

KNN is not memory efficient. It becomes very slow as the number of datapoints increases as the model needs to store all the data points. It is computationally heavy because for a single datapoint the algorithm has to calculate the distance of all the datapoints and find the nearest neighbors.

#Model Evaluation:

Accuracy Score:

In [30]:
#accuracy on the training data

X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [31]:
print('Accuracy on the training data:' , training_data_accuracy)

Accuracy on the training data: 0.9504447268106735


In [32]:
#accuracy on the test data

X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [33]:
print('Accuracy on the test data:' , test_data_accuracy)

Accuracy on the test data: 0.9238578680203046
